File: python/code/10-density-spectrum.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.signal import square

# Sampling frequency
fs = 1000

# Duration
t = np.arange(0, 1, 1 / fs)

# Frequency of square wave
f = 10

# Generate square wave
x = square(2 * np.pi * f * t)

# FFT
X = np.fft.fft(x)

# Power spectrum
power = np.abs(X) ** 2 / len(X)

# Frequency
freq = np.fft.fftfreq(len(x), 1 / fs)

# Positive frequency
positive = freq >= 0

# Plot
plt.figure(figsize=(10, 5))

plt.plot(
    freq[positive],
    power[positive]
)

plt.title("Power Density Spectrum of a Square Wave")
plt.xlabel("Frequency (Hz)")
plt.ylabel("Power")
plt.grid(True)

plt.show()

File: matlab/code/10-density-spectrum.m
clc;
clear;
close all;

% Sampling frequency
Fs = 1000;

% Duration
T = 1;

% Time vector
t = 0:1/Fs:T-1/Fs;

% Square wave frequency
f0 = 10;

% Generate square wave
x = square(2*pi*f0*t);

% Number of samples
N = length(x);

% FFT
X = fft(x);

% Power Density Spectrum
P = (abs(X).^2) / (Fs*N);

% Frequency vector
f = (0:N-1) * Fs/N;

% Plot square wave
figure;
plot(t, x);
grid on;
xlabel('Time (s)');
ylabel('Amplitude');
title('Square Wave');
xlim([0 0.2]);

% Plot Power Density Spectrum
figure;
plot(f(1:N/2), P(1:N/2));
grid on;
xlabel('Frequency (Hz)');
ylabel('Power Density');
title('Power Density Spectrum of Square Wave');
xlim([0 100]);